packages/protocol/src/utils/StreamPartitionID.ts

```diff
@@ -0,0 +1,154 @@
+        this.key =`${this.id}${separator}${this.partition}`
+        this.cachedSplit = [this.id, this.partition]
+        // prevent all mutation
+        Object.freeze(this)
+        Object.freeze(this.cachedSplit)
+    }
+
+    /**
+     * Throws if data is invalid.
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new Error('SPID validation failed: id must be non-empty string')
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new Error('SPID validation failed: paritition must be an integer >= 0 ')
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDish): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = (this.constructor as typeof SPID).from(other)
+        } catch (_err) {
+            // not equal if not valid
+            return false
+        }
+
+        // check if matches both id + partition
+        return (
+            this.id === otherSpid.id
+            && this.partition === otherSpid.partition
+        )
+    }
+
+    /**
+     * Convert to SPID if possible.
+     */
+    static from(spidish: SPIDish): SPID {
+        // return spid if already spid
+        if (spidish instanceof SPID) {
+            return spidish
+        }
+
+        // convert from string
+        if (typeof spidish === 'string') {
+            return SPID.fromString(spidish)
+        } else if (spidish && typeof spidish === 'object') {
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamId = spidish.streamId || spidish.id
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamPartition = spidish.streamPartition || spidish.partition
+            return SPID.from(`${streamId}${this.SEPARATOR}${streamPartition}`)
+        } else {
+            // TODO: add more conversions?
+            throw new Error(`SPID validation failed, input is malformed: ${spidish}`)
+        }
+    }
+
+    /**
+     * Convert String to SPID if possible.
+     */
+    static fromString(spidString: string): SPID {
+        const [id, parititionStr] = spidString.split(this.SEPARATOR)
+        return new SPID(id, Number.parseFloat(parititionStr) || 0)
+    }
+
+    /**
+     * String representation of SPID id + partition
+     */
+    toString(): string {
+        return this.key
+    }
+
+    /**
+     * Alias of toString.
+     */
+    toKey(): string {
+        return this.key
+    }
+
+    /**
+     * Returns an [id, partition] tuple.
+     * Useful for destructuring.
+     * e.g.
+     * ```js
+     * const [streamId, streamPartition] = spid.split()
+     * ```
+     */
+    split(): [string, number] {
```

Comment:
I'm skeptical whether caching the split provides more benefits vs. cost. How often is `split()`  actually used? Put another way: what proportion of `SPID` objects created would have their `split` method invoked more than once?

Reply:
what "cost" though? 

Idea is that since these SPID things are to be used everywhere, ideally the whole class has minimal runtime cost  compared to not using the class, specifically, makes no unnecessary allocations. I'd agree it's premature optimisation if the implementation was remotely complicated, but it seems trivial to just make it do no allocations so why not.

Reply:
Run-time cost:
Given a single instance of this class
- If `split()` is never invoked, we created an array in constructor that was never used.
- If `split()` is invoked once, well then it does get used but is it any better than
```
split(): [string, number] {
    return [this.id, this.partition]
}
```
Conceptual cost:
I'd argue that the snippet above is conceptually simpler than having an extra field to keep track of. (Albeit that field is simple, yes.)

So the way I see it, is that the case where this provides benefit > cost, is that there is reasonably many usages patterns in the code where this method gets called more than once for a single instance?

Reply:
> If split() is never invoked, we created an array in constructor that was never used.

That's fine, it's not the memory I'm trying to optimise for about it's the cost to allocate & GC it. Young generation allocation + GC is fast, generally not something to worry about but seems unnecessary to do any allocation when we can precompute the array upfront. 

> If split() is invoked once, well then it does get used but is it any better …

Intention is that this doesn't require an allocation just to do destructuring:

```js
function onEveryMessage(spid) {
  const [streamId, streamPartition] = spid.split()
  // …
}
```

I was thinking this may be the common usage pattern, but thinking about it now, if everything is using SPIDs then perhaps it won't be necessary and SPIDs could be treated as opaque.

Reply:
Say I create a `SPID` and never use `split` on it? Wouldn't that cause an extra allocation & GC cost that was unnecessary? Are you saying that if you do this allocation in the constructor it's cheaper?

Reply:
removed split